mesonbuild/backend/ninjabackend.py

```diff
@@ -233,17 +233,16 @@ def generate_target(self, target, outfile):
         if isinstance(target, build.Jar):
             self.generate_jar_target(target, outfile)
             return
-        if 'rust' in self.environment.coredata.compilers.keys() and self.has_rust(target):
+        if 'rust' in target.compilers.keys():
             self.generate_rust_target(target, outfile)
             return
-        if 'cs' in self.environment.coredata.compilers.keys() and self.has_cs(target):
+        if 'cs' in target.compilers.keys():
             self.generate_cs_target(target, outfile)
             return
-        if 'vala' in self.environment.coredata.compilers.keys() and self.has_vala(target):
-            vc = self.environment.coredata.compilers['vala']
-            vala_output_files = self.generate_vala_compile(vc, target, outfile)
+        if 'vala' in target.compilers.keys():
```

Comment:
Might be a bit nitpicky, but since vala is the only one that "falls through", I think it should be the last one. Maybe even with a comment about it.